trillian/integration/hammer.go

```diff
@@ -447,17 +602,51 @@ func (s *hammerState) addPreChain(ctx context.Context) error {
 	}
 	submitted.leafHash = sha256.Sum256(append([]byte{ct.TreeLeafPrefix}, submitted.leafData...))
 	s.pending.tryAppendCert(time.Now(), s.cfg.MMD, &submitted)
-	glog.V(3).Infof("%s: Uploaded pre-cert has leaf-hash %x", s.cfg.LogCfg.Prefix, submitted.leafHash)
+	glog.V(3).Infof("%s: Uploaded %s pre-cert has leaf-hash %x", s.cfg.LogCfg.Prefix, choice, submitted.leafHash)
 	return nil
 }
 
 func (s *hammerState) addPreChainInvalid(ctx context.Context) error {
-	// Invalid because it's a cert chain, not a pre-cert chain.
-	prechain, err := makeCertChain(s.cfg.LeafChain, s.cfg.LeafCert, s.cfg.CACert, s.cfg.Signer, s.notAfter)
-	if err != nil {
-		return fmt.Errorf("failed to make fresh pre-cert: %v", err)
+	choices := []Choice{EmptyChain, CertNotPrecert, NoChainToRoot, UnparsableCert, SignatureNotCorrect}
+	choice := choices[rand.Intn(len(choices))]
+
+	var err error
+	var prechain []ct.ASN1Cert
+	switch choice {
+	case EmptyChain:
+	case CertNotPrecert:
+		prechain, err = makeCertChain(s.cfg.LeafChain, s.cfg.LeafCert, s.cfg.CACert, s.cfg.Signer, s.notAfter)
+		if err != nil {
+			return fmt.Errorf("failed to make pre-chain(%s): %v", choice, err)
+		}
+	case NoChainToRoot:
```

Comment:
Add similar comment here about what is being removed from the chain for this case.

Reply:
Done.